packages/core/src/controllers/ConvertController.ts

```diff
@@ -0,0 +1,664 @@
+    )
+    const adjustedTotalSourceCostUSD = totalSourceCostUSD.plus(_gasPriceInUSD)
+    const effectivePricePerTargetToken = adjustedTotalSourceCostUSD.dividedBy(
+      toTokenAmount.toFixed(4)
+    )
+
+    const priceImpact = effectivePricePerTargetToken
+      .minus(state.toTokenPriceInUSD)
+      .dividedBy(state.toTokenPriceInUSD)
+      .multipliedBy(100)
+
+    return priceImpact.toNumber()
+  },
+
+  calculateMaxSlippage() {
+    const slippageToleranceDecimal = NumberUtil.bigNumber(state.slippage).dividedBy(100)
+    const maxSlippageAmount = NumberUtil.multiply(state.sourceTokenAmount, slippageToleranceDecimal)
+
+    return maxSlippageAmount.toNumber()
+  },
+
+  async convertTokens() {
+    const { sourceTokenAddress, toTokenAddress } = this.getParams()
+
+    if (!sourceTokenAddress || !toTokenAddress) {
+      return
+    }
+
+    await this.makeChecks()
+  },
+
+  async makeChecks() {
+    const { toTokenDecimals, toTokenAddress } = this.getParams()
+
+    if (!toTokenDecimals || !toTokenAddress) {
+      return
+    }
+
+    state.loading = true
+    const transaction = await this.getTransaction()
+    this.setTransactionDetails(transaction)
+    state.loading = false
+  },
+
+  async getTransaction() {
+    const { fromAddress, sourceTokenAddress, sourceTokenAmount, sourceTokenDecimals } =
+      this.getParams()
+
+    if (
+      !sourceTokenAddress ||
+      !sourceTokenAmount ||
+      parseFloat(sourceTokenAmount) === 0 ||
+      !sourceTokenDecimals
+    ) {
+      return null
+    }
+
+    const hasAllowance = await ConvertApiController.checkConvertAllowance({
+      fromAddress,
+      sourceTokenAddress,
+      sourceTokenAmount,
+      sourceTokenDecimals
+    })
+
+    let transaction: TransactionParams | null = null
+
+    if (hasAllowance) {
+      state.approvalTransaction = undefined
+      transaction = await this.createConvert()
+      state.convertTransaction = transaction || undefined
+    } else {
+      state.convertTransaction = undefined
+      transaction = await this.createTokenAllowance()
+      state.approvalTransaction = transaction
+    }
+
+    return transaction
+  },
+
+  getToAmount() {
+    const { sourceTokenDecimals } = this.getParams()
+    const decimals = sourceTokenDecimals || 18
+    const multiplyer = 10 ** decimals
+
+    const toTokenConvertedAmount =
+      state.sourceTokenPriceInUSD && state.toTokenPriceInUSD && state.sourceTokenAmount
+        ? NumberUtil.bigNumber(state.sourceTokenAmount)
+            .multipliedBy(state.sourceTokenPriceInUSD)
+            .dividedBy(state.toTokenPriceInUSD)
+        : NumberUtil.bigNumber(0)
+
+    // return toTokenConvertedAmount with 18 decimals
+    return toTokenConvertedAmount.multipliedBy(multiplyer).toString()
+  },
+
+  async createTokenAllowance() {
+    const { fromAddress, sourceTokenAddress } = this.getParams()
+
+    if (!sourceTokenAddress) {
+      throw new Error('>>> createTokenAllowance - No source token address found.')
```

Comment:
Remove throw, instead this should be handled in UI, users shouldn't be able to call this function.